### PR TITLE
Cosoul explore

### PR DESCRIPTION
**Pull request by CryptoGraffe**
### <samp>🤖 Generated by Copilot at 1ee27a6</samp>

This pull request adds a new `profiles_public` view to the database and the GraphQL schema, which exposes public profile information of users. It also removes the `cosoul` feature flag and the related code from the frontend and backend, which disables the CoSoul feature. It also adds a new CoSoul explore page to the frontend, which allows the user to browse the CoSoul items with different orderings.

### <samp>🤖 Generated by Copilot at 1ee27a6</samp>

> _We added a view `profiles_public`_
> _To query the users with GraphQL logic_
> _We also removed_
> _The `cosoul` feature we loved_
> _And made the CoSoul explore page more dynamic_

### <samp>🤖 Generated by Copilot at 1ee27a6</samp>

*  Add a new `profiles_public` view to the database and the Hasura GraphQL engine, and expose it as a GraphQL type with select permissions for anonymous and authenticated users ([link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-412dbd6aa74a0169e4120157039da14d29381e4a3e1cc6551f78afc1616b8e65R1-R20), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-db7a5522001579de7a816ecd82e2a3547b48414c9ae1e552fc0dd24ccac28290R1-R9), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-ca82769adfb1c46558b7227a28ad5929194c8be675ff8867c1a4d7bbf98b67cfR1-R7), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-4147b15d5a605e64a3d76bd68b59506ec8886da36d27df21dbf794443820377fR38), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-8435c41f5069c32f84a110adf7ca73963708315c24f4aabdb0285e87471676ebR11518-R11591), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-8435c41f5069c32f84a110adf7ca73963708315c24f4aabdb0285e87471676ebR13026-R13055), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-8435c41f5069c32f84a110adf7ca73963708315c24f4aabdb0285e87471676ebR15761-R15810), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-59431741d41d2d93c621efe127df0776fc0f78ea9bab64b60cf7418e9f01f060R3816-R3837), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-59431741d41d2d93c621efe127df0776fc0f78ea9bab64b60cf7418e9f01f060R4706-R4714), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-59431741d41d2d93c621efe127df0776fc0f78ea9bab64b60cf7418e9f01f060R7269-R7273), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-59431741d41d2d93c621efe127df0776fc0f78ea9bab64b60cf7418e9f01f060R7340), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-59431741d41d2d93c621efe127df0776fc0f78ea9bab64b60cf7418e9f01f060R7544-R7545))
* Add a new object relationship `profile_public` to the `cosouls` table and the `cosouls` GraphQL type, linking the `address` column of the `cosouls` table to the `address` column of the `profiles_public` view, and allow querying, filtering, and ordering the `profile_public` field of the `cosouls` type ([link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-3837acf731f554c94063419d1b6e3b165007c982ac40c9d93b10416b78fd9532L14-R32), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-8435c41f5069c32f84a110adf7ca73963708315c24f4aabdb0285e87471676ebR4352-R4356), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-8435c41f5069c32f84a110adf7ca73963708315c24f4aabdb0285e87471676ebR4375), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-8435c41f5069c32f84a110adf7ca73963708315c24f4aabdb0285e87471676ebR4391), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-59431741d41d2d93c621efe127df0776fc0f78ea9bab64b60cf7418e9f01f060R1384), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-59431741d41d2d93c621efe127df0776fc0f78ea9bab64b60cf7418e9f01f060R1396), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-59431741d41d2d93c621efe127df0776fc0f78ea9bab64b60cf7418e9f01f060R6469))
* Modify the `auth` middleware function for the Hasura GraphQL engine to allow anonymous users to access the GraphQL endpoint with the role `anon-user`, and to check if the user profile exists in the database before granting the role `user` ([link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-597723888c5385bc188e3ba62a1b6032fda30b4c4bfdb964a2c4364deff98ae0L48-R56), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-597723888c5385bc188e3ba62a1b6032fda30b4c4bfdb964a2c4364deff98ae0R83-R87))
* Remove the `cosoul` feature flag from the `config/features` module and the frontend code, and disable the CoSoul feature in the frontend application ([link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-5c118d9192f4171e7f6592b2205521832975d0666a8673815188b73c9644088aL4), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-5c118d9192f4171e7f6592b2205521832975d0666a8673815188b73c9644088aL15), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-d1cac1e147670959ee5e559069bfc06a6d275b0d913887dd995cd97ba7c360fcL1), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-d1cac1e147670959ee5e559069bfc06a6d275b0d913887dd995cd97ba7c360fcL29-R28), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-a9d33a03447e2b96e70bddf1ebde47791709f7845cfffb8c05a3ee3dacf0348fL6), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-a9d33a03447e2b96e70bddf1ebde47791709f7845cfffb8c05a3ee3dacf0348fL39-L41), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-b43ef01230c6952eb46ff957b2ed7655a981448947dfe5db28cd2e373a648e35L7), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-b43ef01230c6952eb46ff957b2ed7655a981448947dfe5db28cd2e373a648e35L37-L39), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-07a5e546156e72d3af4da968af148298858d1943242859c44c28aea5e4b3c5b5L7), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-07a5e546156e72d3af4da968af148298858d1943242859c44c28aea5e4b3c5b5L46-L49), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-14c17f2206802b962d016d8a767ac1e68950e89f32310c7371d04c790ed7138cL13), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-14c17f2206802b962d016d8a767ac1e68950e89f32310c7371d04c790ed7138cL219-R249))
* Add a new `CoSoulExplorePage` component and a new page to the frontend application that allows the user to explore the CoSoul items with different orderings ([link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-ced0aa6673e9e0b4da48550955daaf19521135613c041dc97534a87f514635e9R1-R52))
* Rename the `CoSoulGalleryPage` component to `DebugCoSoulGalleryPage`, and simplify the code by always rendering the button and the animation ([link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-d1cac1e147670959ee5e559069bfc06a6d275b0d913887dd995cd97ba7c360fcL29-R28), [link](https://github.com/coordinape/coordinape/pull/2355/files?diff=unified&w=0#diff-14c17f2206802b962d016d8a767ac1e68950e89f32310c7371d04c790ed7138cL219-R249))

### CODE REVIEW

File: api/hasura/auth.ts [hunk 1]
```diff
@@ -45,6 +45,11 @@ export default async function handler(req: VercelRequest, res: VercelResponse) {
       return;
     }
 
+    if (!req.headers?.authorization) {
+      return res.status(200).json({
+        'X-Hasura-Role': 'anon-user',
```
Comment (crabsinger) [inline api/hasura/auth.ts:50]: Maybe setup some per-ip rate limiting for anon-user role?

https://hasura.io/learn/graphql/hasura-advanced/security/4-rate-limit/

File: hardhat/contracts/coordinape-protocol
Comment (crabsinger) [inline hardhat/contracts/coordinape-protocol:1]: umm, probably out of date? i don tthink this should be in here?

File: hasura/migrations/default/1696915436967_create_view_profiles_public/down.sql [hunk 2]
```diff
@@ -0,0 +1,9 @@
+-- Could not auto-generate a down migration.
```
Comment (crabsinger) [inline hasura/migrations/default/1696915436967_create_view_profiles_public/down.sql:1]:
```suggestion
DROP VIEW IF EXISTS "public"."profiles_public";
```

File: src/pages/CoSoulExplorePage/CoSoulItem.tsx [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        <Box
+          css={{
+            width: '100%',
+            aspectRatio: '1 / 1',
+          }}
+        >
+          <Image
+            css={{
+              width: '100%',
+              aspectRatio: '1 / 1' /*, top: 0, left: 0, position: 'absolute'*/,
```
Comment (crabsinger) [inline src/pages/CoSoulExplorePage/CoSoulItem.tsx:32]: remove comment

File: src/pages/CoSoulExplorePage/CoSoulItem.tsx [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+              flexGrow: 0,
+              flexShrink: 1,
+            }}
+          >
+            <Avatar
+              path={cosoul.profile_public?.avatar}
+              size="small"
+              name={cosoul.profile_public?.name ?? 'New'}
+            />
+            {/*TODO: if the name is long here this doesn't properly shrink */}
```
Comment (crabsinger) [inline src/pages/CoSoulExplorePage/CoSoulItem.tsx:59]: Maybe use this and find the length that fits right?

```ts
function abbreviateString(str: string, maxLength: number): string {
  if (str.length <= maxLength) {
    return str;
  }
  return str.substring(0, maxLength - 3) + '...';
}
```

File: src/pages/CoSoulExplorePage/useInfiniteCoSouls.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  onSettled?: () => void
+) => {
+  return useInfiniteQuery(
+    queryKey,
+    ({ pageParam = 0 }) => fetchCoSouls(where, orderBy, pageParam),
+    {
+      getNextPageParam: (lastPage, allPages) => {
+        return lastPage.length == 0 ? undefined : allPages.length;
+      },
+      // onSuccess: data => {
```
Comment (crabsinger) [inline src/pages/CoSoulExplorePage/useInfiniteCoSouls.ts:25]:
```suggestion
```

File: src/routes/paths.ts [hunk 6]
```diff
@@ -79,6 +79,7 @@ export const paths = {
   cosoulArt: (tokenId: string) => `/cosoul/art/${tokenId}`,
   cosoulImage: (tokenId: string) => `/cosoul/image/${tokenId}`,
   cosoulGallery: '/cosoul/gallery',
```
Comment (crabsinger) [inline src/routes/paths.ts:81]: should `cosoul/gallery` be live on prod???

File: api/hasura/auth.ts (hunk 1 above)
Comment (CryptoGraffe) [reply api/hasura/auth.ts:50]: YAH THAT WOULD BE SICK

File: hasura/migrations/default/1696915436967_create_view_profiles_public/down.sql (hunk 2 above)
Comment (CryptoGraffe) [reply hasura/migrations/default/1696915436967_create_view_profiles_public/down.sql:1]: ok

File: src/routes/paths.ts (hunk 6 above)
Comment (CryptoGraffe) [reply src/routes/paths.ts:81]: the debug thing? idk prob not @wingfeatherwave

File: src/pages/CoSoulExplorePage/CoSoulItem.tsx (hunk 4 above)
Comment (CryptoGraffe) [reply src/pages/CoSoulExplorePage/CoSoulItem.tsx:59]: did it

**Review by crabsinger — APPROVED** [review]

**Pull request merged**